// graphene-pk11/build/objects/certs/cert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pkcs11 = require("pkcs11js");
var storage_1 = require("../storage");
var CertificateType;
(function (CertificateType) {
    CertificateType[CertificateType["X_509"] = pkcs11.CKC_X_509] = "X_509";
    CertificateType[CertificateType["X_509_ATTR_CERT"] = pkcs11.CKC_X_509_ATTR_CERT] = "X_509_ATTR_CERT";
    CertificateType[CertificateType["WTLS"] = pkcs11.CKC_WTLS] = "WTLS";
})(CertificateType = exports.CertificateType || (exports.CertificateType = {}));
var CertificateCategory;
(function (CertificateCategory) {
    CertificateCategory[CertificateCategory["Unspecified"] = 0] = "Unspecified";
    CertificateCategory[CertificateCategory["TokenUser"] = 1] = "TokenUser";
    CertificateCategory[CertificateCategory["Authority"] = 2] = "Authority";
    CertificateCategory[CertificateCategory["OtherEntity"] = 3] = "OtherEntity";
})(CertificateCategory = exports.CertificateCategory || (exports.CertificateCategory = {}));
var Certificate = (function (_super) {
    tslib_1.__extends(Certificate, _super);
    function Certificate() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(Certificate.prototype, "type", {
        get: function () {
            return this.get("certType");
        },
        set: function (v) {
            this.set("certType", v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Certificate.prototype, "trusted", {
        get: function () {
            return this.get("trusted");
        },
        set: function (v) {
            this.set("trusted", v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Certificate.prototype, "category", {
        get: function () {
            return this.get("certCategory");
        },
        set: function (v) {
            this.set("certCategory", v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Certificate.prototype, "checkValue", {
        get: function () {
            return this.get("checkValue");
        },
        set: function (v) {
            this.set("checkValue", v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Certificate.prototype, "startDate", {
        get: function () {
            return this.get("startDate");
        },
        set: function (v) {
            this.set("startDate", v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Certificate.prototype, "endDate", {
        get: function () {
            return this.get("endDate");
        },
        set: function (v) {
            this.set("endDate", v);
        },
        enumerable: true,
        configurable: true
    });
    return Certificate;
}(storage_1.Storage));
exports.Certificate = Certificate;
