// graphene-pk11/build/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pkcs11 = require("pkcs11js");
var core = require("./core");
var slot = require("./slot");
var Module = (function (_super) {
    tslib_1.__extends(Module, _super);
    function Module(lib) {
        var _this = _super.call(this, lib) || this;
        _this.libFile = "";
        _this.libName = "";
        return _this;
    }
    Module.prototype.getInfo = function () {
        var info = this.lib.C_GetInfo();
        this.cryptokiVersion = info.cryptokiVersion;
        this.manufacturerID = core.removePadding(info.manufacturerID);
        this.libraryDescription = core.removePadding(info.libraryDescription);
        this.flags = info.flags;
        this.libraryVersion = info.libraryVersion;
    };
    Module.prototype.initialize = function () {
        this.lib.C_Initialize();
        this.getInfo();
    };
    Module.prototype.finalize = function () {
        this.lib.C_Finalize();
    };
    Module.prototype.getSlots = function (index, tokenPresent) {
        if (tokenPresent === void 0) { tokenPresent = true; }
        if (!core.isEmpty(index) && core.isBoolean(index)) {
            tokenPresent = index;
        }
        var arr = this.lib.C_GetSlotList(tokenPresent);
        var col = new slot.SlotCollection(arr, this, this.lib);
        if (core.isNumber(index)) {
            return col.items(index);
        }
        return col;
    };
    Module.load = function (libFile, libName) {
        var lib = new pkcs11.PKCS11();
        lib.load(libFile);
        var module = new Module(lib);
        module.libFile = libFile;
        module.libName = libName || libFile;
        return module;
    };
    return Module;
}(core.BaseObject));
exports.Module = Module;
