// graphene-pk11/build/object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pkcs11 = require("pkcs11js");
var core = require("./core");
var template_1 = require("./template");
var ObjectClass;
(function (ObjectClass) {
    ObjectClass[ObjectClass["DATA"] = pkcs11.CKO_DATA] = "DATA";
    ObjectClass[ObjectClass["CERTIFICATE"] = pkcs11.CKO_CERTIFICATE] = "CERTIFICATE";
    ObjectClass[ObjectClass["PUBLIC_KEY"] = pkcs11.CKO_PUBLIC_KEY] = "PUBLIC_KEY";
    ObjectClass[ObjectClass["PRIVATE_KEY"] = pkcs11.CKO_PRIVATE_KEY] = "PRIVATE_KEY";
    ObjectClass[ObjectClass["SECRET_KEY"] = pkcs11.CKO_SECRET_KEY] = "SECRET_KEY";
    ObjectClass[ObjectClass["HW_FEATURE"] = pkcs11.CKO_HW_FEATURE] = "HW_FEATURE";
    ObjectClass[ObjectClass["DOMAIN_PARAMETERS"] = pkcs11.CKO_DOMAIN_PARAMETERS] = "DOMAIN_PARAMETERS";
    ObjectClass[ObjectClass["MECHANISM"] = pkcs11.CKO_MECHANISM] = "MECHANISM";
    ObjectClass[ObjectClass["OTP_KEY"] = pkcs11.CKO_OTP_KEY] = "OTP_KEY";
})(ObjectClass = exports.ObjectClass || (exports.ObjectClass = {}));
var SessionObject = (function (_super) {
    tslib_1.__extends(SessionObject, _super);
    function SessionObject(handle, session, lib) {
        var _this = this;
        if (handle instanceof SessionObject) {
            var obj = handle;
            _this = _super.call(this, obj.handle, obj.lib) || this;
            _this.session = obj.session;
        }
        else {
            _this = _super.call(this, handle, lib) || this;
            _this.session = session;
        }
        return _this;
    }
    Object.defineProperty(SessionObject.prototype, "size", {
        get: function () {
            return this.lib.C_GetObjectSize(this.session.handle, this.handle);
        },
        enumerable: true,
        configurable: true
    });
    SessionObject.prototype.copy = function (template) {
        var tmpl = template_1.Template.toPkcs11(template);
        var hObject = this.lib.C_CopyObject(this.session.handle, this.handle, tmpl);
        return new SessionObject(hObject, this.session, this.lib);
    };
    SessionObject.prototype.destroy = function () {
        this.lib.C_DestroyObject(this.session.handle, this.handle);
    };
    SessionObject.prototype.getAttribute = function (attrs) {
        var _attrs;
        if (typeof attrs === "string") {
            _attrs = {};
            _attrs[attrs] = null;
        }
        else {
            _attrs = attrs;
        }
        var tmpl = template_1.Template.toPkcs11(_attrs);
        tmpl = this.lib.C_GetAttributeValue(this.session.handle, this.handle, tmpl);
        if (typeof attrs === "string") {
            return template_1.Template.fromPkcs11(tmpl)[attrs];
        }
        return template_1.Template.fromPkcs11(tmpl);
    };
    SessionObject.prototype.setAttribute = function (attrs, value) {
        if (core.isString(attrs)) {
            var tmp = {};
            tmp[attrs] = value;
            attrs = tmp;
        }
        var tmpl = template_1.Template.toPkcs11(attrs);
        this.lib.C_SetAttributeValue(this.session.handle, this.handle, tmpl);
    };
    SessionObject.prototype.get = function (name) {
        var tmpl = {};
        tmpl[name] = null;
        return this.getAttribute(tmpl)[name];
    };
    SessionObject.prototype.set = function (name, value) {
        var tmpl = {};
        tmpl[name] = value;
        this.setAttribute(tmpl);
    };
    Object.defineProperty(SessionObject.prototype, "class", {
        get: function () {
            return this.get("class");
        },
        set: function (v) {
            this.set("class", v);
        },
        enumerable: true,
        configurable: true
    });
    SessionObject.prototype.toType = function () {
        var c = this.class;
        switch (c) {
            case ObjectClass.DATA:
                return new objects.Data(this);
            case ObjectClass.DOMAIN_PARAMETERS:
                return new objects.DomainParameters(this);
            case ObjectClass.CERTIFICATE:
                var cert = new objects.Certificate(this);
                var t = cert.type;
                switch (t) {
                    case objects.CertificateType.X_509:
                        return new objects.X509Certificate(this);
                    case objects.CertificateType.WTLS:
                        return new objects.WtlsCertificate(this);
                    case objects.CertificateType.X_509_ATTR_CERT:
                        return new objects.AttributeCertificate(this);
                    default:
                        throw new Error("Unknown certificate (CKC_?) type '" + t + "'");
                }
            case ObjectClass.PRIVATE_KEY:
                return new objects.PrivateKey(this);
            case ObjectClass.PUBLIC_KEY:
                return new objects.PublicKey(this);
            case ObjectClass.SECRET_KEY:
                return new objects.SecretKey(this);
            case ObjectClass.HW_FEATURE:
            case ObjectClass.OTP_KEY:
                throw new Error("Type converter for " + ObjectClass[c] + " is not implemented");
            default:
                throw new Error("Unknown session object (CKO_?) type '" + c + "'");
        }
    };
    return SessionObject;
}(core.HandleObject));
exports.SessionObject = SessionObject;
var SessionObjectCollection = (function (_super) {
    tslib_1.__extends(SessionObjectCollection, _super);
    function SessionObjectCollection(items, session, lib, classType) {
        if (classType === void 0) { classType = SessionObject; }
        var _this = _super.call(this, items, lib, classType) || this;
        _this.session = session;
        return _this;
    }
    SessionObjectCollection.prototype.items = function (index) {
        return new SessionObject(this.items_[index], this.session, this.lib);
    };
    return SessionObjectCollection;
}(core.Collection));
exports.SessionObjectCollection = SessionObjectCollection;
var objects = require("./objects/common");
tslib_1.__exportStar(require("./objects/common"), exports);
tslib_1.__exportStar(require("./keys/common"), exports);
